pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AstDisambiguationPass.java

```diff
@@ -0,0 +1,469 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+
+package net.sourceforge.pmd.lang.java.ast;
+
+import static net.sourceforge.pmd.lang.java.symbols.table.internal.SemanticChecksLogger.CANNOT_RESOLVE_AMBIGUOUS_NAME;
+import static net.sourceforge.pmd.lang.java.symbols.table.internal.SemanticChecksLogger.CANNOT_RESOLVE_MEMBER;
+
+import java.util.Iterator;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.lang.ast.NodeStream;
+import net.sourceforge.pmd.lang.ast.impl.javacc.JavaccToken;
+import net.sourceforge.pmd.lang.java.internal.JavaAstProcessor;
+import net.sourceforge.pmd.lang.java.symbols.JClassSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JFieldSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JTypeDeclSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JVariableSymbol;
+import net.sourceforge.pmd.lang.java.symbols.table.JSymbolTable;
+import net.sourceforge.pmd.lang.java.symbols.table.ResolveResult;
+import net.sourceforge.pmd.lang.java.symbols.table.internal.SemanticChecksLogger;
+
+/**
+ * This implements name disambiguation following <a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.5.2">JLS§6.5.2</a>.
+ * (see also <a href="https://docs.oracle.com/javase/specs/jls/se13/html/jls-6.html#jls-6.4.2">JLS§6.4.2 - Obscuring</a>)
+ *
+ * <p>Currently disambiguation of package vs type name is fully implemented,
+ * with the following limitations: TODO
+ * - field accesses are not checked to be legal (and their symbol is not
+ * resolved).
+ * - inherited members are not considered. Same thing happens in the current
+ * symbol table. Since that is so, visibility/accessibility is not handled either.
+ * See bottom of file for some test cases.
+ *
+ * This is because we don't have full access to types yet. This is the next
+ * step.
+ */
+public final class AstDisambiguationPass {
```

Comment:
That's in the public package `lang.java.ast`, but this is a implementation, isn't? Should it be rather in a internal package?

Reply:
Hmm that's right. Well it uses a lot of package-private API so probably the best way is to make it package-private and add a hook to it in InternalApiBridge